src/journal/journald-server.c

```diff
@@ -2525,6 +2568,13 @@ int server_init(Server *s, const char *namespace) {
 
         server_parse_config_file(s);
 
+        server_load_credentials(s);
+
+        /* only open a forwarding socket if a forwarding address has been set
+         * and we are in the main namespace */
+        if (s->forward_to_socket.sockaddr.sa.sa_family != AF_MAX && !s->namespace)
+                server_open_forward_socket(s);
```

Comment:
hmm, i'd drop this here I must say.

I think it would be nicer to simply try to call server_open_forward_socket() from server_forward_socket() when the socket fd is not yet set up. If it works, then use it, if it doesn't try on the next datagram again.

And conversely, if the forwarding fails with some error in sendmsg() I think we should close the fd again.

Why? since for IP and stuff failures might happen and are transient, and we should just keep going making the best of it. It's just more robust, and means we can work when we can work.

This applies even to AF_VSOCK: there are multiple backends to AV_VSOCK and there might be some we only load late via /etc/modules-load.d/, and that makes sense journald migh tninitially not be able to use AF_VSOCK but later might be.

Reply:
How to handle a connection dropping etc has been something on my mind for a while so I like that this solves it at the same time :)